### PR TITLE
Fix the sequence lengths mismatch

### PR DESCRIPTION
**Pull request by tingkai-zhang**
The sequence lengths of input and output should match each other.
Because `do_layer_norm=False` in self._scalar_mix, line 87 should just pass `layer_activations` according to api documentation.

### CODE REVIEW

**Comment by tingkai-zhang** [comment]
> the current behavior is correct (or, if it's not, you haven't convinced me of it)
> 
> The point of a `TokenEmbedder` is to produce a vector embedding per text-input token. In most cases you're right that the input to the TokenEmbedder = output of the TokenIndexer is also one indexed vector per text-input token. However, for things like byte pair encoding (which is what we have here) the output of the token indexer is one indexed vector per byte-pair, which is (usually) more than the number of tokens. In which case we shouldn't expect the output of the TokenEmbedder to have the same sequence length as its input.
> 
> It sort of seems like you want to use the OpenAITransformer but not as a `TokenEmbedder`? In that case maybe just use `OpenAITransformer` directly?

Thank you for replying!

I think you are talking about `OpenaiTransformerBytePairIndexer` 's job. It tokenize text sentence which makes the input and output has different sequence length.(because word to be tokenized BPE representation, which has more tokens)

I am using `OpenAITransformer ` as `TokenEmbedder`

```
    openai_transformer = OpenaiTransformer(model_path='openai-transformer-lm-2018.07.23.tar.gz',
                                           n_ctx=512,
                                           n_special=2)

    token_embedders = {"openai_transformer_embedder": 
   OpenaiTransformerEmbedder(openai_transformer)}

```

First, the idea of taking the last byte-pair hidden state to represent the token doesn't make sense to me because that will lose the token(sentence) information, which will make it not be able to reproduce the experiment result in original paper. It is not the idea in original implementation neither([tensorflow](https://github.com/openai/finetune-transformer-lm/blob/ae7e86b55254339eba3929f786b04e82602cd51d/train.py#L172) and [pytorch](https://github.com/huggingface/pytorch-openai-transformer-lm/blob/eafc28abdfadfa0732f03a0fc65805c5bfb2ffe7/model_pytorch.py#L170)).



Second, the input **is** one index per byte-pair because you have to use `OpenaiTransformerBytePairIndexer`, which tokenize the sentence to byte-type pair and index them.


```
    tokenizer = WordTokenizer(word_splitter=OpenAISplitter(),
                              start_tokens=[START_TOKEN],
                              end_tokens=[END_TOKEN])

    token_indexer = OpenaiTransformerBytePairIndexer(n_ctx=512,
                                                     model_path='openai-transformer-lm-2018.07.23.tar.gz',
                                                     tokens_to_add=[START_TOKEN, END_TOKEN]
                                                     )
```

Third, I can still to find the index of last token(here i mean BPE represetation),  because the last token is the special token ['_classify_`], whose BPE has length 1. But returning last BPE of each word( i mean, word like 'apple' not 'app', 'le</w>') is not correct. Consequently, the mask returned in [openai byte pair indexer](https://github.com/allenai/allennlp/blob/68cbfb800fbf173fd6a1c097c6dd73530065ae1d/allennlp/data/token_indexers/openai_transformer_byte_pair_indexer.py#L228) is not correct.
Check #2020  


As a conclusion, despite the truncation things in [openai_token_embedder](https://github.com/allenai/allennlp/blob/68cbfb800fbf173fd6a1c097c6dd73530065ae1d/allennlp/modules/token_embedders/openai_transformer_embedder.py#L98)
 and all the padding things,

**Comment by tingkai-zhang** [comment]
Initially, all i want is the input and output of Text field Embedder has same number of timestamps, which what the original implementation does.
After reading the code, i think your allennlp implementation is not correct(to me).

For logical level,  taking only last byte-pair representation for each word will lose information.
For code level, it  makes downstream tasks difficult to expand because lengths mismatch.



##########################################################
I am not sure if you have reproduced the same result in the paper. The result may be correct, but that's because the original implementation only takes the last byte pair representation of each **sentence**. And here, we can also get the last byte pair representation of each **sentence**.

**Comment by matt-gardner** [comment]
The entire point of the `OpenaiTransformerEmbedder` is to get back a sequence of vectors for each token from the byte-pair sequence.  So, if you want something that doesn't do that, `OpenaiTransformerEmbedder` is not what you are looking for.  The `OpenaiTransformerEmbedder` does very little on top of the `OpenaiTransformer` - you probably should just be using that directly.
